### PR TITLE
ObservedGeneration and update block

### LINKED ISSUE #1274: k8ssandra-operator shouldn't modify the cassdc objects on operator upgrade
k8ssandra-operator in spec changes to the cassdc objects which will instantly trigger a rolling restart when the operator is upgraded.
As one operator instance may manage lots of clusters, we don't want upgrades to trigger a rolling restart storm.

```[tasklist]
### Tasks
- [ ] If the Generation of the K8ssandraCluster is the same as the ObservedGeneration we don't update the CassDC objects in Kubernetes (instead, we mark the K8ssandraCluster as requiring refresh). Metadata updates would still be applied.
- [ ] A condition should be placed on the object to indicate that a restart is pending (something as "RestartRequired" condition)
- [ ] Restart can rely on a new annotation `k8ssandra.io/allow-upgrade=once` (using `k8ssandra.io/allow-upgrade=always` would allow permanently to automatically update the K8ssandraCluster on upgrade). This annotation would be placed by either the users or the rolling restart K8ssandraTask. If this annotation is detected, the CassDc Spec update will be unblocked, the "RestartRequired" condition will be set to false and the annotation will be removed at the end of the reconcile if its value is set to `once`
- [ ] We need to allow reconcile to trigger if this annotation is placed on the objects (k8c). This requires an additional predicate for annotation listeners, but then one more filter. This could be useful outside of this feature also.
```

### PR DESCRIPTION
**Pull request by burmanm**
**What this PR does**:
Adds new status field: ObservedGeneration and ensure it is updated at the end of the reconciliation

**Which issue(s) this PR fixes**:
Fixes #1274 

**Checklist**
- [ ] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Comment by burmanm** [comment]
@olim7t

File: controllers/control/k8ssandratask_controller.go
Comment (olim7t) [inline controllers/control/k8ssandratask_controller.go:213]: [thought] It really bothers me to add an entirely new way to execute a task just for this.
- do we really need a task just to set an annotation on the cluster?
- is there any way we could handle DCs as part of the reconcile loop? That is, if we process a K8ssandraCluster with auto-update=once, then we propagate it to any DC with RequiresUpdate.

File: controllers/k8ssandra/k8ssandracluster_controller_test.go
Comment (burmanm) [reply controllers/k8ssandra/k8ssandracluster_controller_test.go:2707]: Yes, this is just a hack for the testing purpose. The idea is that the new version of k8ssandra-operator would create a different hash (since it wants to create different kind of CassandraDatacenter) which then wouldn't match the old one.

There's no other easy way to cause this in the test without causing Generation to change also (which we don't want in the test).

File: apis/k8ssandra/v1alpha1/constants.go
Comment (burmanm) [reply apis/k8ssandra/v1alpha1/constants.go:76]: That would blow up the PR, since updating cass-operator is not the issue, but the old dependencies in k8ssandra-operator.

So updating cass-operator requires updating Kubernetes dependencies, controller-runtime dependencies and those are not backwards compatible with the current code, so there would be more changes.

To give a scope of the required changes, here are the changes that cass-operator alone required: https://github.com/k8ssandra/cass-operator/commit/c273e274310b0b007178ab214b2e32774f3cf819

**Comment by adejanovski** [comment]
@burmanm , is this ready for another round of review?

**Comment by burmanm** [comment]
Yes

File: controllers/k8ssandra/datacenters.go
Comment (olim7t) [inline controllers/k8ssandra/datacenters.go:160]: [nitpick] the message is wrong, we're trying to set the condition, not the annotation.

File: controllers/k8ssandra/datacenters.go
Comment (olim7t) [inline controllers/k8ssandra/datacenters.go:262]: [issue] If I'm following correctly, we might requeue for a few reconciliation loops while we wait for the task to finish. However, each loop creates its own `generatedName`, which might be different because the UNIX timestamp part has a resolution of one second. With bad timing, we might end up creating multiple tasks.

**Review by olim7t — CHANGES_REQUESTED** [review]
I still have to test this manually, but publishing this review early because I think the task naming issue needs attention.

File: apis/k8ssandra/v1alpha1/k8ssandracluster_types.go
Comment (olim7t) [inline apis/k8ssandra/v1alpha1/k8ssandracluster_types.go:577]: [thought] Since this PR also introduces the new field, maybe we should add this to correctly detect the first operator upgrade:
```suggestion
	return k.Status.ObservedGeneration != 0 && k.Status.ObservedGeneration < k.Generation
```

File: controllers/k8ssandra/datacenters.go
Comment (olim7t) [reply controllers/k8ssandra/datacenters.go:262]: Just to confirm, manual testing confirms that the operator keeps creating refresh tasks in a loop.

```yaml
apiVersion: k8ssandra.io/v1alpha1
kind: K8ssandraCluster
metadata:
  namespace: k8ssandra-operator
  name: demo
spec:
  cassandra:
    serverVersion: 4.1.0
    datacenters:
      - metadata:
          name: dc1
        size: 1
        racks:
          - name: rack1
        storageConfig:
          cassandraDataVolumeClaimSpec:
            storageClassName: standard
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 100Mi
```

Simulate a cass-operator upgrade by setting the condition on the `CassandraDatacenter` manually:
```zsh
# Find out the index of the condition in the array. If this is different than 10,
# you need to change it in the next command
kubectl get CassandraDatacenter dc1 -n k8ssandra-operator -o json \
  | jq '.status.conditions | map(.type == "RequiresUpdate") | index(true)'

kubectl patch CassandraDatacenter dc1 -n k8ssandra-operator \
  --subresource status \
  --type='json' \
  -p='[{"op": "replace", "path": "/status/conditions/10/status", "value": "True"}]'
```

Set the annotation on the cluster:
```zsh
kubectl annotate K8ssandraCluster demo -nk8ssandra-operator \
  k8ssandra.io/autoupdate-spec=once
```

Expected: a single `K8ssandraTask` is created.
Actual: a new one is created every ~20s.

**Review by olim7t — APPROVED** [review]
I pushed commits to fix a few remaining issues. This is good to go as far as k8ssandra-operator is concerned.

However I'm noticing an issue with the refresh `CassandraTask`: it doesn't seem to ever be marked as finished (and as a consequence neither does the `K8ssandraTask`). It doesn't impact the new behavior we're introducing here, the conditions and annotations are correctly reset.

I'll test on cass-operator alone and file an issue if necessary.

**Comment by olim7t** [comment]
> However I'm noticing an issue with the refresh `CassandraTask`: it doesn't seem to ever be marked as finished (and as a consequence neither does the `K8ssandraTask`).

OK, this might be a side-effect of my testing method: I was artificially setting the `RequiresUpdate=true` condition on the `CassandraDatacenter` to simulate a cass-operator upgrade, but refreshing the DC did not produce any actual changes to the stateful set. It looks like something in the way the refresh is implemented causes the task to not be marked as completed in that case. This is not something that will happen in real life, so I think we can safely ignore it.

**Comment by adejanovski** [comment]
Given we found some bugs (or side effects) in cass-operator's implementation of this feature, I think we should first solve those and re-evaluate how it's done in this PR.
Also, and that's probably something we should have done from the beginning, I think we should make this an opt in feature rather than the default behavior.

**Comment by adejanovski** [comment]
@burmanm, as discussed together, could you add some docs before we merge this? We'd need to explain users what the new behavior looks like and how to unblock a spec update on operator upgrade.
Thanks!

File: docs/content/en/install/upgrade/_index.md
Comment (rzvoncek) [inline docs/content/en/install/upgrade/_index.md:13]: `on inconvenient time` -> `at inconvenient time`

**Review by rzvoncek — APPROVED** [review]
Reviewed the docs update (`docs/content/en/install/upgrade/_index.md`).
There's one small improvement I found, but I couldn't get GH to let me propose a commit.

File: docs/content/en/install/upgrade/_index.md
Comment (burmanm) [reply docs/content/en/install/upgrade/_index.md:13]: I think the correct form is "at an inconvenient time" ;)

Comment (olim7t) [reply docs/content/en/install/upgrade/_index.md]: "at an inconvenient time" sounds like the best wording to me.

File: apis/k8ssandra/v1alpha1/k8ssandracluster_types.go
Comment (burmanm) [reply apis/k8ssandra/v1alpha1/k8ssandracluster_types.go]: Turns out, this causes a bug. ObservedGeneration is 0 in new clusters also and that's perfectly valid - so we have to allow update in that case, otherwise we'll never update the CassandraDatacenters. I think we'll just have to accept that older clusters won't benefit from restart protection the first time.

**Pull request merged**
